reco_utils/recommender/rbm/rbm.py

```diff
@@ -0,0 +1,946 @@
+
+        Main component of the algo; once instantiate, it generates the computational graph and performs
+        model training
+
+        Args:
+            Xtr: the user/affinity matrix for the train set
+
+        Returns:
+            trained_param: the trained values for the weight matrix and the visible/hidden bias vectors
+            msre : the mean square root error per training epoch
+            precision: the mean average precision over the entire dataset
+
+        Baic mechanics:
+
+
+        '''
+
+        #keep the position of the items in the train set so that they can be optionally exluded from recommendation
+        self.seen_mask = np.not_equal(xtr,0)
+
+        #start timing the methos
+        self.time()
+
+        self.r_= xtr.max() #defines the rating scale, e.g. 1 to 5
+        m, self.Nv_ = xtr.shape #dimension of the input: m= N_users, Nv= N_items
+
+        num_minibatches = int(m / self.minibatch) #number of minibatches
+
+        tf.reset_default_graph()
+        #----------------------Initialize all parameters----------------
+
+        log.info("Creating the computational graph")
+
+        #create the visible units placeholder
+        self.placeholder()
+
+        self.batch_size_ = tf.placeholder(tf.int64)
+
+        #Create data pipeline
+        self.dataset = tf.data.Dataset.from_tensor_slices(self.vu)
+        self.dataset = self.dataset.shuffle(buffer_size= 50, reshuffle_each_iteration=True, seed = 1) #randomize the batch
+        self.dataset = self.dataset.batch(batch_size= self.batch_size_).repeat()
+
+        self.iter = self.dataset.make_initializable_iterator()
+        self.v = self.iter.get_next()
+
+        #initialize Network paramters
+        self.init_parameters()
+
+        #--------------Sampling protocol for Gibbs sampling-----------------------------------
+        self.k=1 #initialize the G_sampling step
+        self.l=0 #initialize epoch_sample index
+        #-------------------------Main algo---------------------------
+
+        self.G_sampling() #returns the sampled value of the visible units
+
+        obj = self.Losses(self.v) #objective function
+        rate = self.alpha/self.minibatch  #learning rate rescaled by the batch size
+
+        #Instantiate the momentum optimizer
+        opt = tf.contrib.optimizer_v2.AdamOptimizer(learning_rate = rate).minimize(loss= obj)
+
+        pvh, vp = self.infere() #sample the value of the visible units given the hidden. Also returns  the related probabilities
+
+        if self.with_metrics_:#if true (default) returns evaluation metrics
+            Mse_train = [] #Lists to collect the metrics across epochs
+            #Metrics
+            Mserr  = self.msr_error(self.v_k)
+            Clacc  = self.accuracy(self.v_k)
+
+        if self.save_model_: #save the model to file
+            saver = tf.train.Saver()
+
+        init_g = tf.global_variables_initializer() #Initialize all variables in the graph
+
+        #Config GPU memory
```

Comment:
I would give the option to overwrite this, instead of defining it here. That way the user can control these parameters

Reply:
Sorry, I am not sure I understood what you meant here.

Reply:
This code:
```
Config = tf.ConfigProto(log_device_placement = True, allow_soft_placement=True)
Config.gpu_options.per_process_gpu_memory_fraction = 0.5
```
is configuring parts of the process, right now if I want to use more than 50% of the GPU memory I can't do it, because it is fixed in the code. If instead we add these parameters as inputs, the programmer can choose the usage of the GPU programatically.

Reply:
ok got it. That was one "experiment", thanks for bringing it to my attention. The reason is that I experience GPU memory problems when using the 10m dataset. I have also tried to allow TF to increase the memory allocation dynamically but it did not solve the problem. An easy solution would be to move to a DSVM with more memory (currently using the smallest available). I will open an issue if you think it is worth spending some time on this.

Reply:
I changed this back to 

`
Config = tf.ConfigProto(log_device_placement=True, allow_soft_placement=True)
Config.gpu_options.allow_growth = True  # dynamics memory allocation
`
In this way memory allocation is dynamical and the user does not need to modify anything.